README
NAME
    File::Symlink::Atomic - an atomic drop-in replacement for CORE::symlink

VERSION
    version 0.001

SYNOPSIS
        use File::Symlink::Atomic;   # imports replacement symlink
        symlink "target", "name1";   # easy peasy
        symlink "bullseye", "name1"; # now atomic

DESCRIPTION
    Actually creating a symlink is not problematic, but making an existing
    one point at a new target may not be atomic on your system. For example,
    on Linux, the system does "unlink" and then "symlink". In between, no
    symlink exists. If something goes wrong, you're left with nothing.

    In your shell, you probably want to do something like:

        mkdir old-target new target # Create your targets
        ln -s old-target link       # Create your initial symlink
        # ln -sf new-target link    # NOT atomic!
        ln -s new-target link-tmp && mv -Tf link-tmp link

    Moving the symlink to the new name makes it atomic, because under the
    hood, the "mv" command does "rename", which is guaranteed to be atomic
    by POSIX
    <http://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html>.

    File::Symlink::Atomic attempts to do the same thing in Perl what the
    command shown above does for your shell.

FUNCTIONS
  symlink OLDFILE,NEWFILE
    Creates a new filename symbolically linked to the old filename. Returns
    1 for success, 0 otherwise. This drop-in replacement for "CORE::symlink"
    creates a symlink with a temporary name, then renames it to the name you
    requested - this ensures that if a symlink by the requested name already
    existed, then its target is updated atomically.

CAVEATS
    This module is not guaranteed to be portable. I have no idea what this
    will do on any platform other than Linux. Feel free to run the test
    suite to find out!

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <https://metacpan.org/module/File::Symlink::Atomic/>.

SOURCE
    The development version is on github at
    <http://github.com/doherty/File-Symlink-Atomic> and may be cloned from
    <git://github.com/doherty/File-Symlink-Atomic.git>

BUGS AND LIMITATIONS
    You can make new bug reports, and view existing ones, through the web
    interface at <https://github.com/doherty/File-Symlink-Atomic/issues>.

AUTHOR
    Mike Doherty <doherty@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Mike Doherty.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl module {{$dist->name}}

{{$NEXT}}

0.002     2012-06-12
  - Bail unless it is a Unix-ey platform [GH #2]

0.001     2012-06-10
  - Initial revision


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "File::Find" => 0,
    "Module::Build" => "0.3601",
    "Test::More" => 0,
    "Test::Requires" => 0
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "an atomic drop-in replacement for CORE::symlink",
  "dist_author" => [
    "Mike Doherty <doherty\@cpan.org>"
  ],
  "dist_name" => "File-Symlink-Atomic",
  "dist_version" => "0.001",
  "license" => "perl",
  "module_name" => "File::Symlink::Atomic",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Exporter" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/File/Symlink/Atomic.pm
package File::Symlink::Atomic;
use strict;
use warnings;
# ABSTRACT: an atomic drop-in replacement for CORE::symlink
# VERSION

use File::Temp;
use File::Spec;
use Exporter qw(import);
our @EXPORT = qw(symlink); ## no critic(Modules::ProhibitAutomaticExportation)

=head1 SYNOPSIS

    use File::Symlink::Atomic;   # imports replacement symlink
    symlink "target", "name1";   # easy peasy
    symlink "bullseye", "name1"; # now atomic

=head1 DESCRIPTION

Actually creating a symlink is not problematic, but making an existing one point
at a new target may not be atomic on your system. For example, on Linux, the
system does C<unlink> and then C<symlink>. In between, no symlink exists. If
something goes wrong, you're left with nothing.

In your shell, you probably want to do something like:

    mkdir old-target new target # Create your targets
    ln -s old-target link       # Create your initial symlink
    # ln -sf new-target link    # NOT atomic!
    ln -s new-target link-tmp && mv -Tf link-tmp link

Moving the symlink to the new name makes it atomic, because under the hood, the
C<mv> command does C<rename>, which is guaranteed to be atomic by
L<POSIX|http://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html>.

B<File::Symlink::Atomic> attempts to do the same thing in Perl what the command
shown above does for your shell.

=head1 FUNCTIONS

=head2 symlink OLDFILE,NEWFILE

Creates a new filename symbolically linked to the old filename.
Returns C<1> for success, C<0> otherwise. This drop-in replacement
for C<CORE::symlink> creates a symlink with a temporary name, then
renames it to the name you requested - this ensures that if a
symlink by the requested name already existed, then its target is
updated atomically.

=cut
    
sub symlink($$) { ## no critic(Subroutines::ProhibitSubroutinePrototypes Subroutines::ProhibitBuiltinHomonyms)
    my $symlink_target = shift;
    my $symlink_name   = shift;

    my ($volume, $dirs, $file) = File::Spec->splitpath($symlink_name);
    my $template = File::Spec->catpath($volume, $dirs, ".$file.$$.XXXXXX");
    my $tmp_symlink_name;
    ATTEMPT:
    for (1..10) { # try 10 times
        $tmp_symlink_name = mktemp($template);
        symlink $symlink_target, $tmp_symlink_name and last ATTEMPT;
    }
    return 0 unless -l $tmp_symlink_name; # wtf?
    
    rename $tmp_symlink_name, $symlink_name or return 0; # should be atomic
    return 1;
}

1;

=head1 CAVEATS

This module is B<not> guaranteed to be portable. I have no idea what this will
do on any platform other than Linux. Feel free to run the test suite to find out!

=cut


Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "an atomic drop-in replacement for CORE::symlink",
  "AUTHOR" => "Mike Doherty <doherty\@cpan.org>",
  "BUILD_REQUIRES" => {
    "File::Find" => 0,
    "Module::Build" => "0.3601",
    "Test::More" => 0,
    "Test::Requires" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "File-Symlink-Atomic",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "File::Symlink::Atomic",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



